### PR TITLE
C#: Fix extraction of qualified delegate calls

### PR DESCRIPTION
**Pull request by hvitved**
This PR fixes an extractor bug where we would incorrectly determine the qualifier of a delegate call when it was a qualified expression. For example, in
```csharp
this.LambdaField(0)
```
where `LambdaField` is a field holding a lambda, we would consider `this` to be the qualifier of the delegate call, instead of `this.LambdaField`.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs [hunk 1]
```diff
@@ -33,7 +37,7 @@ protected override void PopulateExpression(TextWriter trapFile)
             var target = TargetSymbol;
             switch (Syntax.Expression)
             {
-                case MemberAccessExpressionSyntax memberAccess:
+                case MemberAccessExpressionSyntax memberAccess when Kind == ExprKind.METHOD_INVOCATION || IsEventDelegateCall() || IsExplicitDelegateInvokeCall():
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs:40]: When is this `when` condition not the same as `Kind == ExprKind.METHOD_INVOCATION || Kind = ExprKind.DELEGATE_INVOCATION`?

File: csharp/ql/test/library-tests/delegates/delegates.cs [hunk 2]
```diff
@@ -83,4 +83,17 @@ static void Main()
...
     }
 
+    class E
+    {
+        Action<int> Field;
+        Action<int> Property { get; set; }
+
+        void M()
+        {
+            this.Field(0);
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/delegates/delegates.cs:93]: Do we have the same problem with function pointers? (`unsafe delegate*<int, void> Field2;`)

**Review by michaelnebel — APPROVED** [review]
Looks good to me!
How/why did you discover this problem?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs (hunk 1 above)
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/Invocation.cs:40]: Exactly when the member being accessed is not an event or a direct `.Invoke`, i.e., when it is a field or a property.

File: csharp/ql/test/library-tests/delegates/delegates.cs (hunk 2 above)
Comment (hvitved) [reply csharp/ql/test/library-tests/delegates/delegates.cs:95]: We did, yes, but that is also fixed by this PR. I have added additional tests to witness.

**Pull request merged**